# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentParent.cpp
# Commit: 9213b6bda67a
# Full Hash: 9213b6bda67a689644f442f7b98902c5a5184c74
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-09-24 09:46:13
# Regressor Bug: 1725572
# File Overlap Count: 1
# Description:
#   Bug 1725572 - Part 2: Keep BrowsingContext alive until every process has acknowledged the discard, r=farre
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D125897
# ==============================================================================

diff -r 6bb38c7309e5 -r 9213b6bda67a dom/ipc/ContentParent.cpp
--- a/dom/ipc/ContentParent.cpp	Thu Sep 23 18:54:31 2021 +0000
+++ b/dom/ipc/ContentParent.cpp	Thu Sep 23 18:54:31 2021 +0000
@@ -6663,15 +6663,19 @@
 }
 
 mozilla::ipc::IPCResult ContentParent::RecvDiscardBrowsingContext(
-    const MaybeDiscarded<BrowsingContext>& aContext,
+    const MaybeDiscarded<BrowsingContext>& aContext, bool aDoDiscard,
     DiscardBrowsingContextResolver&& aResolve) {
-  if (!aContext.IsNullOrDiscarded()) {
-    RefPtr<CanonicalBrowsingContext> context = aContext.get_canonical();
-    if (!CheckBrowsingContextEmbedder(context, "discard")) {
-      return IPC_FAIL(this, "Illegal Discard attempt");
+  if (CanonicalBrowsingContext* context =
+          CanonicalBrowsingContext::Cast(aContext.GetMaybeDiscarded())) {
+    if (aDoDiscard && !context->IsDiscarded()) {
+      if (!CheckBrowsingContextEmbedder(context, "discard")) {
+        return IPC_FAIL(this, "Illegal Discard attempt");
+      }
+
+      context->Detach(/* aFromIPC */ true);
     }
-
-    context->Detach(/* aFromIPC */ true);
+    context->AddFinalDiscardListener(aResolve);
+    return IPC_OK();
   }
 
   // Resolve the promise, as we've received and handled the message. This will